File: SmartHomeTEC_API/Models/UsageLog.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using System.Text.Json.Serialization;

namespace SmartHomeTEC_API.Models
{
    [Table("UsageLog")]
    public class UsageLog
    {
        [Key]
        [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
        public int? LogID { get; set; }

        [Required]
        [RegularExpression(@"^\d{4}-\d{2}-\d{2}$", ErrorMessage = "StartDate debe tener el formato 'YYYY-MM-DD'.")]
        public required string StartDate { get; set; } // Formato: YYYY-MM-DD

        [Required]
        [RegularExpression(@"^\d{2}:\d{2}$", ErrorMessage = "StartTime debe tener el formato 'HH:mm'.")]
        public required string StartTime { get; set; } // Formato: HH:mm

        [RegularExpression(@"^\d{4}-\d{2}-\d{2}$", ErrorMessage = "EndDate debe tener el formato 'YYYY-MM-DD'.")]
        public string? EndDate { get; set; } // Formato: YYYY-MM-DD

        [RegularExpression(@"^\d{2}:\d{2}$", ErrorMessage = "EndTime debe tener el formato 'HH:mm'.")]
        public string? EndTime { get; set; } // Formato: HH:mm

        public string? TotalHours { get; set; } // Calculado en base a StartDate/StartTime y EndDate/EndTime

        [Required]
        [EmailAddress]
        [ForeignKey("Client")]
        public required string ClientEmail { get; set; }

        [Required]
        [ForeignKey("AssignedDevice")]
        public required int AssignedID { get; set; }

        // Propiedades de Navegación:
        [JsonIgnore]
        public Client? Client { get; set; }

        [JsonIgnore]
        public AssignedDevice? AssignedDevice { get; set; }
    }
}

File: smart_home_tec_app/lib/JSONmodels/usage_log.dart
import 'dart:convert';

UsageLog usageLogFromJson(String str) => UsageLog.fromJson(json.decode(str));

String usageLogToJson(UsageLog data) => json.encode(data.toJson());

class UsageLog {
    final int? logId;
    final String startDate;
    final String startTime;
    final String? endDate;
    final String? endTime;
    final String? totalHours;
    final String clientEmail;
    final int assignedId;

    UsageLog({
        this.logId,
        required this.startDate,
        required this.startTime,
        this.endDate,
        this.endTime,
        this.totalHours,
        required this.clientEmail,
        required this.assignedId,
    });

    factory UsageLog.fromJson(Map<String, dynamic> json) => UsageLog(
        logId: json["logID"],
        startDate: json["startDate"],
        startTime: json["startTime"],
        endDate: json["endDate"],
        endTime: json["endTime"],
        totalHours: json["totalHours"],
        clientEmail: json["clientEmail"],
        assignedId: json["assignedID"],
    );

    Map<String, dynamic> toJson() => {
        "logID": logId,
        "startDate": startDate,
        "startTime": startTime,
        "endDate": endDate,
        "endTime": endTime,
        "totalHours": totalHours,
        "clientEmail": clientEmail,
        "assignedID": assignedId,
    };
}
